src/Microsoft.ML.Data/Commands/TrainCommand.cs

```diff
@@ -252,77 +252,27 @@ public static string MatchNameOrDefaultOrNull(IExceptionContext ectx, ISchema sc
             ch.CheckValueOrNull(validData);
             ch.CheckValueOrNull(inpPredictor);
 
-            var trainerRmd = trainer as ITrainer<RoleMappedData>;
-            if (trainerRmd == null)
-                throw ch.ExceptUserArg(nameof(TrainCommand.Arguments.Trainer), "Trainer '{0}' does not accept known training data type", name);
-
-            Action<IChannel, ITrainer, Action<object>, object, object, object> trainCoreAction = TrainCore;
-            IPredictor predictor;
             AddCacheIfWanted(env, ch, trainer, ref data, cacheData);
             ch.Trace("Training");
             if (validData != null)
                 AddCacheIfWanted(env, ch, trainer, ref validData, cacheData);
 
-            var genericExam = trainCoreAction.GetMethodInfo().GetGenericMethodDefinition().MakeGenericMethod(
-                typeof(RoleMappedData),
-                inpPredictor != null ? inpPredictor.GetType() : typeof(IPredictor));
-            Action<RoleMappedData> trainExam = trainerRmd.Train;
-            genericExam.Invoke(null, new object[] { ch, trainerRmd, trainExam, data, validData, inpPredictor });
-
-            ch.Trace("Constructing predictor");
-            predictor = trainerRmd.CreatePredictor();
+            var trainerEx = trainer as ITrainerEx;
+            if (inpPredictor != null && trainerEx?.SupportsIncrementalTraining != true)
+            {
+                ch.Warning("Ignoring " + nameof(TrainCommand.Arguments.InputModelFile) +
+                    ": Trainer does not support incremental training.");
+                inpPredictor = null;
+            }
```

Comment:
Here it gives me a feeling that its caller responsibility to check for this warning. However, it should be the `trainer` that should warn the caller that wrong set of arguments were passed because its the trainer who know which arguments to use and which one to ignore.

Also, I think this path won't be followed in ML.Net so I don't think that this warning will be printed in ML.Net, right? (though I would prefer to see this warning in case I am expecting something else from the trainer)

Reply:
No I don't really think so. A caller that cares about whether the validation set or initialization predictor will be used really ought to check or know that. I don't want it to be the callee's responsilibity, for the reasons mentioned elsewhere.

---
In reply to: [202174443](https://github.com/dotnet/machinelearning/pull/522#discussion_r202174443) [](ancestors = 202174443)

Reply:
ok, make sense.

---
In reply to: [202182888](https://github.com/dotnet/machinelearning/pull/522#discussion_r202182888) [](ancestors = 202182888,202174443)